sw/host/opentitanlib/src/io/emu.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transport::Result;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use thiserror::Error;
+
+/// Error related to the `Emulator` trait. These error messages will be printed in the context of
+/// a TransportError::EmuError, that is "Emulator error: {}".  So including the words "error" or
+/// "Emulator" in texts below will probably be redundant.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum EmuError {
+    #[error("Invalid argument name {0}")]
+    InvalidArgumetName(String),
+    #[error("Argument {0} has invalid value {1}")]
+    InvalidArgumentValue(String, String),
+    #[error("Start failed with cause {0}")]
+    StartFailureCause(String),
+    #[error("Stop failed with cause {0}")]
+    StopFailureCause(String),
+    #[error("Can't restore resource to initial state: {0}")]
+    ResetError(String),
+    #[error("Runtime error {0}")]
+    RuntimeError(String),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum EmuValue {
+    Empty,
+    String(String),
+    FilePath(String),
+    StringList(Vec<String>),
+    FilePathList(Vec<String>),
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
+pub enum EmuState {
+    Off,
+    On,
+    Busy,
+    Error,
+}
+
+/// A trait which represents a `Emulator` instance
+pub trait Emulator {
+    /// Simple function with return `EmuState` representing current state of Emulator instance.
+    fn get_state(&self) -> Result<EmuState>;
+
+    /// Start Emulator with provided arguments.
+    /// Parameter `factory_reset` is a flag that describe if internal state
+    /// and "resources" should be set to its "initial state" before applying
+    /// the changes from the parameter `args`.
+    /// If `factory_reset` is set to true - all internal state and "resources"
+    /// will be set to its "initial state".
+    /// If `factory_reset` is set to false - all internals state will be preserved
+    /// form last run. The "initial state" for most devices means that the content
```

Comment:
Move the remark about "initial state" often meaning all zeroes up to the paragraph above, which describes about the case of factory_reset = true.